### PR TITLE
Compact chat

### PR DESCRIPTION
**Pull request by Alexdoru**
Compacts identical consecutive chat messages together

image

~~Also disabled by default the message "You can only sleep at night" rendering above the hotbar~~

### CODE REVIEW

**Comment by miozune** [comment]
> Also disabled by default the message "You can only sleep at night" rendering above the hotbar

Any reason? It's the same behavior as newer versions of mc

**Comment by Alexdoru** [comment]
> Any reason? It's the same behavior as newer versions of mc

I thought that it was added to prevent chat spamming, since with compact chat there is no spam anymore I tough I would disable it

**Comment by miozune** [comment]
Bed message spam was fixed predate hotbar HUD, https://github.com/GTNewHorizons/Hodgepodge/pull/102. I introduced https://github.com/GTNewHorizons/Hodgepodge/pull/205 because it looks more fancy

**Comment by Alexdoru** [comment]
> Bed message spam was fixed predate hotbar HUD, #102. I introduced #205 because it looks more fancy

It's just my personal opinion, since this pr adds compact chat, I'd prefer to see the bed message in the chat and not above the hotbar, ngl when testing at first I didn't even see the bed message above the hotbar until I went into the git log to find what happened lol.

Feel free to put the setting back the true by default if you want

File: src/main/java/com/mitchej123/hodgepodge/client/chat/ChatHandler.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import net.minecraft.util.IChatComponent;
+
+public class ChatHandler {
+
+    private static int previousCount = 1;
+
+    /**
+     * Returns true if it compacts the message
+     */
+    public static boolean tryCompactMessage(IChatComponent imsg, @SuppressWarnings("rawtypes") List chatLines) {
```
Comment (glowredman) [inline src/main/java/com/mitchej123/hodgepodge/client/chat/ChatHandler.java:15]: Use `List<ChatLine>`

File: src/main/java/com/mitchej123/hodgepodge/mixins/early/minecraft/MixinGuiNewChat_CompactChat.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import org.spongepowered.asm.mixin.injection.callback.LocalCapture;
+
+import com.mitchej123.hodgepodge.client.chat.ChatHandler;
+
+@Mixin(GuiNewChat.class)
+public class MixinGuiNewChat_CompactChat {
+
+    @Shadow
+    @Final
+    private List chatLines;
```
Comment (glowredman) [inline src/main/java/com/mitchej123/hodgepodge/mixins/early/minecraft/MixinGuiNewChat_CompactChat.java:26]: Use `List<ChatLine>`

File: src/main/java/com/mitchej123/hodgepodge/mixins/early/minecraft/MixinGuiNewChat_CompactChat.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+    @Shadow
+    @Final
+    private List field_146253_i; // drawnChatLines
```
Comment (glowredman) [inline src/main/java/com/mitchej123/hodgepodge/mixins/early/minecraft/MixinGuiNewChat_CompactChat.java:30]: Use `List<ChatLine>`

File: src/main/java/com/mitchej123/hodgepodge/mixins/early/minecraft/MixinGuiNewChat_CompactChat.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        this.hodgepodge$deleteMessage = !refresh && ChatHandler.tryCompactMessage(imsg, this.chatLines);
+    }
+
+    @Inject(
+            method = "func_146237_a",
+            at = @At(value = "INVOKE", target = "Lnet/minecraft/client/gui/GuiNewChat;getChatOpen()Z"),
+            locals = LocalCapture.CAPTURE_FAILSOFT)
+    private void hodgepodge$deletePrevious(IChatComponent imsg, int p_146237_2_, int p_146237_3_, boolean refresh,
+            CallbackInfo ci, int k, int l, ChatComponentText chatcomponenttext,
+            @SuppressWarnings("rawtypes") ArrayList arraylist) {
```
Comment (glowredman) [inline src/main/java/com/mitchej123/hodgepodge/mixins/early/minecraft/MixinGuiNewChat_CompactChat.java:47]: Use `ArrayList<IChatComponent>`

**Review by glowredman — CHANGES_REQUESTED** [review]
I agree with @miozune that the current behaviour should be kept.

**Comment by Alexdoru** [comment]
I did not add the types at first because I did not know how it would behave with the source code that doesn't have them

**Comment by Dream-Master** [comment]
@Alexdoru can you resolve the conflict ?

**Review by glowredman — APPROVED** [review]

**Pull request merged**